agent_app_sample_code/utils/get_recursive_character_text_splitter.py

```diff
@@ -0,0 +1,180 @@
+HF_CACHE_DIR = "/tmp/hf_cache/"
+
+
+EMBEDDING_MODELS = {
+    "gte-large-en-v1.5": {
+        "tokenizer": lambda: AutoTokenizer.from_pretrained(
+            "Alibaba-NLP/gte-large-en-v1.5", cache_dir=HF_CACHE_DIR
+        ),
+        "context_window": 8192,
+        "type": "SENTENCE_TRANSFORMER",
+    },
+    "bge-large-en-v1.5": { # FMAPI Per-Token name
+        "tokenizer": lambda: AutoTokenizer.from_pretrained(
+            "BAAI/bge-large-en-v1.5", cache_dir=HF_CACHE_DIR
+        ),
+        "context_window": 512,
+        "type": "SENTENCE_TRANSFORMER",
+    },
+    "bge_large_en_v1_5": { # FMAPI PT name
+        "tokenizer": lambda: AutoTokenizer.from_pretrained(
+            "BAAI/bge-large-en-v1.5", cache_dir=HF_CACHE_DIR
+        ),
+        "context_window": 512,
+        "type": "SENTENCE_TRANSFORMER",
+    },
+    "text-embedding-ada-002": {
+        "context_window": 8192,
+        "tokenizer": lambda: tiktoken.encoding_for_model("text-embedding-ada-002"),
+        "type": "OPENAI",
+    },
+    "text-embedding-3-small": {
+        "context_window": 8192,
+        "tokenizer": lambda: tiktoken.encoding_for_model("text-embedding-3-small"),
+        "type": "OPENAI",
+    },
+    "text-embedding-3-large": {
+        "context_window": 8192,
+        "tokenizer": lambda: tiktoken.encoding_for_model("text-embedding-3-large"),
+        "type": "OPENAI",
+    },
+}
+
+
+def detect_fmapi_embedding_model_type(model_serving_endpoint):
+    """
+    Try to detect the endpoint type and get the embedding config
+    returns (None, None) if not found
+    returns (endpoint_type, embedding_config)
+    """
+    w = WorkspaceClient()
+
+    endpoint_type = None
+    try:
+        llm_endpoint = w.serving_endpoints.get(name=model_serving_endpoint)
+        # external model name
+        endpoint_type = llm_endpoint.config.served_entities[0].external_model.name
+    except Exception as e:
+        try:
+            # FMAPI pay per token
+            endpoint_type = llm_endpoint.config.served_entities[0].foundation_model.name
+        except Exception as e:
+            try: 
+                # FMAPI provisioned throughput
+                endpoint_type = llm_endpoint.config.served_entities[0].name
+                if endpoint_type[-2:][0] == '-':
+                    # remove the version number if present e.g., `bge_large_en_v1_5-2` --> `bge_large_en_v1_5`
+                    endpoint_type = endpoint_type[:-2]
+            except Exception as e:
+                pass
+                
+
+    if endpoint_type is not None:
+        # will return None if not found
+        embedding_config = EMBEDDING_MODELS.get(endpoint_type)
+        if embedding_config is not None:
+            return (endpoint_type, embedding_config)
+        else:
+            return None, None
+    else:
+        return None, None
+
+
+def get_recursive_character_text_splitter(
```

Comment:
Also wondering if we need to maintain this code in the cookbook or can lift/copy from somewhere in Databricks docs, and whether all this config actually has to live in user code

Reply:
core logic for this is in `langchain_text_splitters.RecursiveCharacterTextSplitter`, so this code is just setting up args to call that with. 

i think this code is necessary tho for us to have a runnable parser + VS pipeline which is the first step for most agent/rag workflows.